### DISCUSSION TITLE
Get RouterCollection using RoutingConfigurator

### DISCUSSION THREAD

**Discussion by eerison** [symfony/symfony#54268] [Q&A] [answered]
Hello guys atm I am getting RouterCollection using

```php
<?php

declare(strict_types=1);

namespace crmi\Loader;

use Symfony\Component\Routing\Loader\AnnotationClassLoader;
use Symfony\Component\Routing\Route;

class DefaultAnnotationClassLoader extends AnnotationClassLoader
{
    protected function configureRoute(
        Route $route,
        \ReflectionClass $class,
        \ReflectionMethod $method,
        object $annot
    ): void {
        $route->setDefault('_controller', sprintf('%s::%s', $method->class, $method->name));
        $route->setDefault('_action', $method->name);
        $route->setDefault('_path', $route->getPath());
    }
}
```

```php
//my front controller
$locator = new FileLocator(APPLICATION_ROOT . '/lib');
$loader = new AnnotationDirectoryLoader($locator, new DefaultAnnotationClassLoader());
$routes = $loader->load('controller/Admin/');
$routes->addPrefix('/admin');
$routes->addNamePrefix('admin_');

return $routes;
```

it is working! But I would like to load this from a `routes` folder

e.g:

```php
//config/crmi/routes/attributes.php

<?php

use Symfony\Component\Routing\Loader\Configurator\RoutingConfigurator;

return static function (RoutingConfigurator $routes): void {
    $routes->import(
        // this is added to the beginning of all imported route names
        '../../../lib/controller/Admin/',
        'annotations',
        false,
        // the optional fourth argument is used to exclude some files
        // or subdirectories when loading attributes
        // (the value must be a string or an array of PHP glob patterns)
        // '../../src/Controller/{Debug*Controller.php}'
    )
        ->prefix('/admin')
        ->namePrefix('admin_');
};
```

```php
//my front controller
$locator = new FileLocator(APPLICATION_ROOT . '/config/crmi');
$loader = new AnnotationDirectoryLoader($locator, new DefaultAnnotationClassLoader());

return $loader->load('routes');
```

But it is not working :/
do you have any idea how could I make it work?

**Comment by stof** [comment]
The issue with `$route->setDefault('_path', $route->getPath());` is that you are reading the route path at the time your DefaultAnnotationClassLoader is called, not the final path that will be in the route object once the route loading is completed. The Route object is a mutable object. Adding routes in a collection with a prefix will mutate the path.

**Reply by eerison** [reply]
Hi Stof, thank you for your quickly reply <3

I removed this `_path` (I don't need this anymore)

But I am still facing this issue

```
Controller of type '\crmi\controller\Admin' was not found
```

I guess it is other this :/


Note: my class is like this now

```php
<?php

declare(strict_types=1);

namespace crmi\Loader;

use Symfony\Component\Routing\Loader\AnnotationClassLoader;
use Symfony\Component\Routing\Route;

class DefaultAnnotationClassLoader extends AnnotationClassLoader
{
    protected function configureRoute(
        Route $route,
        \ReflectionClass $class,
        \ReflectionMethod $method,
        object $annot
    ): void {
        $route->setDefault('_controller', sprintf('%s::%s', $method->class, $method->name));
        $route->setDefault('_action', $method->name);
    }
}
```

**Reply by eerison** [reply]
@stof 

the current way it return the list of routes (there are 2 routes)
<img width="879" alt="Screenshot 2024-03-13 at 14 46 34" src="https://github.com/symfony/symfony/assets/6358755/1a6ef839-c652-44da-801d-d356b3486639">

and using RouterConfigurator, it doesn't return any route :/ for this reason it is not found.

<img width="573" alt="Screenshot 2024-03-13 at 14 47 30" src="https://github.com/symfony/symfony/assets/6358755/3a162cd6-eafd-4ae6-bd57-8619cc188f80">

But why it didn't collect any route :'(

**Reply by eerison** [reply]
<img width="879" alt="Screenshot 2024-03-13 at 14 46 34" src="https://github.com/symfony/symfony/assets/6358755/d0d53f12-1430-4a75-b34b-7d2d07c117e1">
it also have 3 resources .... hmmmmm 

Note: it is the first way (without RouterConfigurator)

**Reply by eerison** [reply]
I also saw, that it should be `annotation` :D
https://symfony.com/doc/5.x/routing/custom_route_loader.html#loading-routes

well But it isn't the problem, I am still getting no route :/

**Comment by eerison** [comment]
Ok I guess we are almost there!

it worked like this

```php
$locator = new FileLocator(APPLICATION_ROOT . '/config/crmi');
$resolver = new LoaderResolver([
    new AnnotationDirectoryLoader($locator, new DefaultAnnotationClassLoader()),
    new PhpFileLoader($locator),
]);
$loader = new DelegatingLoader($resolver);

return $loader->load('routes/attributes.php');
```

But I would like to load more the one route file :/ for example

- routes/attributes.php
- routes/redirects.php

**Reply by eerison** [reply, marked as answer]
Ok I imported redirects.php into attributes.php and it worked :)

Note: I will rename attributes.php to routes.php and keep this file as main one, and for others files I just import ;)
